chips/qemu_rv32_virt_chip/src/uart.rs

```diff
@@ -0,0 +1,468 @@
+//! QEMU's memory mapped 16550 UART
```

Comment:
Tock already has a 16550 compatible UART, can you not use that?

Reply:
Huh, that's interesting. It even took me a while to find the file you're referring to. I didn't try to reinvent the wheel, rather parts of this code have already been written >6 months ago collecting dust and so this was developed entirely independent. I think both implementations have some distinct advantages (for instance, this has extensive `register_bitfields!` and less bitmasks, whereas yours uses `register_structs!`) and are in general very similar. I'd suggest I try merging them and putting them in a common place (given that they aren't bound to a specific chip), if that's okay with you? I'd also be integrating read as I'm close to have that working locally.

Reply:
Yep, that sounds good to me

Reply:
Okay, I spent some significant time on getting the SweRVolf running (wouldn't work on Nix and requires a source-compiled Verilator on Debian as the package is too outdated) and merging these two. It's significantly more difficult than I expected (without ending up with a barely usable interface) because while these are very similar, the memory mapped interfaces aren't the same: QEMU presents the (usually exposed on an x86 I/O bus) "registers" as a byte-packed array in memory, whereas the SweRVolf puts them in 32-bit wide memory locations. I solved that issue by treating them all like `u8` registers and having a configurable const-generic based offset between the registers.

... but it still doesn't work. I can't trace down as to why exactly, but also didn't spent much time on debugging yet. I'd say that it might be much more reasonable to keep these separate for now and care about merging them later? In a separate PR we can then also resolve the (for now unanswered) question as to where shared peripherals between chips across vendors are stored.

Reply:
Hmmm... That's annoying that they are treated differently. I agree generics sound like the best way to handle this.

I don't see why we need two 16550 compatible UARTs in Tock though. 16550 compatible UART is pretty common so including one that can be used by multiple chips/boards seems really useful. It seem reasonable to have a "generic" chip with support for the UART.